### PR TITLE
Ensure RazorPages in an area are not route-able through root based pa…

### LINKED ISSUE #7147: Ensure RazorPages in an area are not route-able through root based paths when root directory and area root directory overlap
See https://github.com/aspnet/Mvc/pull/7135#discussion_r157549019 for further discussion

Consider this scenario:

```
/Areas/First.cshtml
/Areas/MyArea/Pages/Second.cshtml
where RootDirectory = /, AreaRootDirectory = /Areas
```

In this case, we want to ensure that the two routes registered are /Areas/First and /MyArea/Second

### PR DESCRIPTION
**Pull request by pranavkm**
…ths when root directory and area root directory overlap

Fixes #7147

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs [hunk 1]
```diff
@@ -63,21 +63,31 @@ public static void PopulateDefaults(PageRouteModel model, string pageRoute, stri
                 return false;
             }
 
-            // Normalize the pages root directory so that it has a 
-            var normalizedPagesRootDirectory = razorPagesOptions.RootDirectory.TrimStart('/');
-            if (!normalizedPagesRootDirectory.EndsWith("/", StringComparison.Ordinal))
+            string normalizedPagesRootDirectory;
+            if (razorPagesOptions.RootDirectory == "/")
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs]: Had to special case empty root "/". I'm going to try and see if I can make a change to StringTokenizer (need the enumerator to expose the string index) and clean this up.

File: build/dependencies.props [hunk 2]
```diff
@@ -62,7 +62,7 @@
     <MicrosoftExtensionsLoggingTestingPackageVersion>2.1.0-preview1-27942</MicrosoftExtensionsLoggingTestingPackageVersion>
     <MicrosoftExtensionsObjectMethodExecutorSourcesPackageVersion>2.1.0-preview1-27942</MicrosoftExtensionsObjectMethodExecutorSourcesPackageVersion>
     <MicrosoftExtensionsOptionsPackageVersion>2.1.0-preview1-27942</MicrosoftExtensionsOptionsPackageVersion>
-    <MicrosoftExtensionsPrimitivesPackageVersion>2.1.0-preview1-27942</MicrosoftExtensionsPrimitivesPackageVersion>
+    <MicrosoftExtensionsPrimitivesPackageVersion>2.1.0-preview1-t005a78e88</MicrosoftExtensionsPrimitivesPackageVersion>
```
Comment (rynowak) [inline build/dependencies.props]: This is a kinda sketch build number, is this should be here?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs [hunk 3]
```diff
@@ -73,13 +73,13 @@ private void EnsureCache()
                 foreach (var viewDescriptor in GetViewDescriptors(_applicationManager))
                 {
                     PageRouteModel model = null;
-                    if (viewDescriptor.RelativePath.StartsWith(rootDirectory, StringComparison.OrdinalIgnoreCase))
+                    if (_pagesOptions.AllowAreas && viewDescriptor.RelativePath.StartsWith(areaRootDirectory, StringComparison.OrdinalIgnoreCase))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: The fact that we're relying on the relationship between arearoot and root here isn't obvious. This deserves a comment that the ordering is significant.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs [hunk 4]
```diff
@@ -56,39 +57,46 @@ public static void PopulateDefaults(PageRouteModel model, string pageRoute, stri
             result = default;
             Debug.Assert(path.StartsWith("/", StringComparison.Ordinal));
 
-            var areaEndIndex = path.IndexOf('/', startIndex: 1);
-            if (areaEndIndex == -1 || areaEndIndex == path.Length)
+            var stringTokenizer = new StringTokenizer(path, PathSeparator);
+            var enumerator = stringTokenizer.GetEnumerator();
+            var next = enumerator.MoveNext();
+            Debug.Assert(next && enumerator.Current.Length == 0, "Paths must start with leading slash");
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs]: I think L62-63 deserve a comment that we're moving 'past' the leading slash

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs [hunk 5]
```diff
@@ -56,39 +57,46 @@ public static void PopulateDefaults(PageRouteModel model, string pageRoute, stri
...
+                var normalizedRootDirectory = new StringSegment(rootDirectory, 1, rootDirectory.Length - 1);
+                var rootDirectoryTokenizer = new StringTokenizer(normalizedRootDirectory, PathSeparator);
+                foreach (var segment in rootDirectoryTokenizer)
+                {
+                    if (!enumerator.MoveNext() || !segment.Equals(enumerator.Current, StringComparison.OrdinalIgnoreCase))
+                    {
+                        logger.UnsupportedAreaPath(razorPagesOptions, path);
+                        return false;
+                    }
+                }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs]: I really have mixed feelings on seeing this code written with the tokenizer. It's much harder to understand

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs]: Fair enough. I can stick to the current code. I'll see if I can document \ clean it up further and simplify the special case of an empty root directory.

**Pull request closed without merging**